overlord/snapstate/handlers.go

```diff
@@ -1061,6 +1061,44 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 
 	// Notify link snap participants about link changes.
 	notifyLinkParticipants(t, snapsup.InstanceName())
+
+	// undo migration if appropriate
+	if snapsup.Flags.Revert {
+		opts, err := getDirMigrationOpts(st, snapst, snapsup)
+		if err != nil {
+			return err
+		}
+
+		newInfo, err := readInfo(snapsup.InstanceName(), snapsup.SideInfo, errorOnBroken)
+		if err != nil {
+			return err
+		}
+
+		action := triggeredMigration(oldInfo.Base, newInfo.Base, opts)
+
+		switch action {
+		case revertFull:
+			if err := m.backend.UndoHideSnapData(snapsup.InstanceName()); err != nil {
+				return err
+			}
+
+			snapsup.UndidHiddenMigration = true
+			snapsup.DisableExposedHome = true
+		case revertHidden:
+			if err := m.backend.UndoHideSnapData(snapsup.InstanceName()); err != nil {
+				return err
+			}
+
+			snapsup.UndidHiddenMigration = true
+		case disableHome:
+			snapsup.DisableExposedHome = true
```

Comment:
That's a fair concern but I think this is one of those situations where if that happens then we're probably in a bad state and that won't help us much. The migration flags in SnapSetup are false initially and they're only set in places that do or undo migrations. If there were leftovers in them it's because we did something we shouldn't have and resetting them removes the information that we did it but not the effects of migration operation itself.

Reply:
OK!